File: packages/kernel/src/animation.ts
// Animation sampling helpers for KHR_interactivity's animation control
// operations (animation/start, animation/stop, animation/stopAt) and the
// /animations/{n}/extensions/KHR_interactivity/* virtual pointers. Extracted
// verbatim (behavior-preserving) from packages/runtime/src/interpreter.ts so
// both the interpreter and the compiled runtime-lib engine can reuse the
// exact same sampling/timeline math — see docs/design/ir-and-transpiler.md's
// M3 refactor note. Environment-neutral: reads the glTF JSON document and an
// optional DataView over the GLB binary chunk, both handed in explicitly
// (no dependency on a "runtime" or "engine" shape).
import { quatSlerp } from "./math.js";

export type AccessorHost = {
  gltf: any;
  // Binary buffer chunk (GLB BIN) when available; sampling degrades
  // gracefully when it is null (accessor reads return null).
  glbBin: DataView | null;
};

const ACCESSOR_COMPONENT_COUNTS: Record<string, number> = {
  SCALAR: 1, VEC2: 2, VEC3: 3, VEC4: 4, MAT2: 4, MAT3: 9, MAT4: 16
};

// Decodes an accessor from the binary chunk as an array of per-element number
// arrays. Only tightly-packed float32 data is supported, which covers
// animation samplers. Returns null when no binary chunk is available.
export function readAccessorElements(host: AccessorHost, accessorIndex: number): number[][] | null {
  const accessor = host.gltf?.accessors?.[accessorIndex];
  const bin = host.glbBin;
  if (!accessor || !bin || accessor.componentType !== 5126) {
    return null;
  }
  const componentCount = ACCESSOR_COMPONENT_COUNTS[accessor.type] ?? 1;
  const view = host.gltf?.bufferViews?.[accessor.bufferView ?? -1];
  if (!view) {
    return null;
  }
  const byteOffset = (view.byteOffset ?? 0) + (accessor.byteOffset ?? 0);
  const end = byteOffset + accessor.count * componentCount * 4;
  if (end > bin.byteLength) {
    return null;
  }
  const out: number[][] = [];
  for (let i = 0; i < accessor.count; i += 1) {
    const element: number[] = [];
    for (let j = 0; j < componentCount; j += 1) {
      element.push(bin.getFloat32(byteOffset + (i * componentCount + j) * 4, true));
    }
    out.push(element);
  }
  return out;
}

export function getAnimationTimeRange(host: AccessorHost, animationIndex: number): { min: number; max: number } {
  const animation = host.gltf?.animations?.[animationIndex];
  if (!animation) {
    return { min: NaN, max: NaN };
  }
  let min = Infinity;
  let max = -Infinity;
  const usedSamplers = new Set<number>((animation.channels ?? []).map((channel: any) => channel.sampler));
  for (const samplerIndex of usedSamplers) {
    const sampler = animation.samplers?.[samplerIndex];
    const input = host.gltf?.accessors?.[sampler?.input ?? -1];
    if (!input) {
      continue;
    }
    if (Array.isArray(input.min)) {
      min = Math.min(min, input.min[0]);
    }
    if (Array.isArray(input.max)) {
      max = Math.max(max, input.max[0]);
    }
  }
  if (!Number.isFinite(min) || !Number.isFinite(max)) {
    return { min: NaN, max: NaN };
  }
  return { min, max };
}

// Maps a requested timestamp on the infinite timeline to the effective
// timestamp within the animation data range [0, T] (KHR_interactivity §4.6).
export function effectiveAnimationTime(requested: number, maxTime: number): number {
  if (!maxTime) {
    return 0;
  }
  const s = requested > 0 ? Math.ceil((requested - maxTime) / maxTime) : Math.floor(requested / maxTime);
  return requested - s * maxTime;
}

export type ApplyAnimationResult = { t: number; wroteAny: boolean };

// Samples animation `animationIndex` at virtual-timeline position `requested`
// and writes the resulting channel values straight into `host.gltf`'s node
// objects (translation/rotation/scale/weights), calling `onChannelWrite` for
// each written channel (pointer path + sampled value) so a caller can notify
// its own dirty/observer hooks. Returns null when the animation doesn't
// exist; otherwise the effective sample time `t` and whether any channel was
// actually written (no binary chunk / no accessor data => false).
export function applyAnimationAt(
  host: AccessorHost,
  animationIndex: number,
  requested: number,
  onChannelWrite?: (pointer: string, value: number[]) => void
): ApplyAnimationResult | null {
  const animation = host.gltf?.animations?.[animationIndex];
  if (!animation) {
    return null;
  }
  const { max } = getAnimationTimeRange(host, animationIndex);
  const t = effectiveAnimationTime(requested, Number.isFinite(max) ? max : 0);
  let wroteAny = false;
  for (const channel of animation.channels ?? []) {
    const sampler = animation.samplers?.[channel.sampler];
    const target = channel.target;
    if (!sampler || !target || target.node === undefined) {
      continue;
    }
    const times = readAccessorElements(host, sampler.input)?.map((item) => item[0]);
    const values = readAccessorElements(host, sampler.output);
    if (!times || !values || times.length === 0) {
      continue;
    }
    const interpolation = sampler.interpolation ?? "LINEAR";
    const stride = interpolation === "CUBICSPLINE" ? 3 : 1;
    let sampled: number[];
    if (t <= times[0]) {
      sampled = values[0 * stride + (stride === 3 ? 1 : 0)];
    } else if (t >= times[times.length - 1]) {
      sampled = values[(times.length - 1) * stride + (stride === 3 ? 1 : 0)];
    } else {
      let k = 0;
      while (k + 1 < times.length && times[k + 1] < t) {
        k += 1;
      }
      const t0 = times[k];
      const t1 = times[k + 1];
      const u = t1 > t0 ? (t - t0) / (t1 - t0) : 0;
      const v0 = values[k * stride + (stride === 3 ? 1 : 0)];
      const v1 = values[(k + 1) * stride + (stride === 3 ? 1 : 0)];
      if (interpolation === "STEP") {
        sampled = v0;
      } else if (target.path === "rotation") {
        sampled = quatSlerp(v0, v1, u);
      } else {
        sampled = v0.map((item, index) => item + (v1[index] - item) * u);
      }
    }
    const node = host.gltf?.nodes?.[target.node];
    if (!node) {
      continue;
    }
    if (target.path === "weights") {
      node.weights = sampled;
    } else {
      node[target.path] = sampled;
    }
    onChannelWrite?.(`/nodes/${target.node}/${target.path}`, sampled);
    wroteAny = true;
  }
  return { t, wroteAny };
}

File: packages/runtime-lua/src/lua/animation.lua
-- Animation sampling helpers, transcribed from packages/kernel/src/
-- animation.ts. `host` is a Lua table `{ gltf = <decoded JSON>, glbBin =
-- <binary Lua string or nil> }`. All glTF-JSON indices (accessor index,
-- animation index, node index, ...) arrive here 0-based (that's the glTF
-- spec's own convention, baked into the document itself) — every place
-- below that indexes into a decoded-JSON array adds 1 for Lua's 1-based
-- tables; byte offsets into the binary chunk stay 0-based arithmetic (raw
-- byte math, not a Lua array index) until the final `string.unpack` call,
-- whose own position argument is 1-based (Lua string-library convention),
-- hence the `+ 1` there specifically.

local ACCESSOR_COMPONENT_COUNTS = { SCALAR = 1, VEC2 = 2, VEC3 = 3, VEC4 = 4, MAT2 = 4, MAT3 = 9, MAT4 = 16 }

-- Decodes an accessor from the binary chunk as an array of per-element
-- Lua-array (1-based) number lists. Only tightly-packed float32 data is
-- supported (covers animation samplers). Returns nil when unavailable.
function anim_readAccessorElements(host, accessorIndex)
  local accessors = host.gltf and host.gltf.accessors
  local accessor = accessors and accessors[accessorIndex + 1]
  local bin = host.glbBin
  if not accessor or not bin or accessor.componentType ~= 5126 then
    return nil
  end
  local componentCount = ACCESSOR_COMPONENT_COUNTS[accessor.type] or 1
  local bufferViews = host.gltf.bufferViews
  local view = bufferViews and bufferViews[(accessor.bufferView or 0) + 1]
  if not view then
    return nil
  end
  local byteOffset = (view.byteOffset or 0) + (accessor.byteOffset or 0)
  local count = accessor.count
  local finish = byteOffset + count * componentCount * 4
  if finish > #bin then
    return nil
  end
  local out = {}
  for i = 0, count - 1 do
    local element = {}
    for j = 0, componentCount - 1 do
      local byte = byteOffset + (i * componentCount + j) * 4
      element[j + 1] = string.unpack("<f", bin, byte + 1)
    end
    out[i + 1] = element
  end
  return out
end

function anim_getAnimationTimeRange(host, animationIndex)
  local animations = host.gltf and host.gltf.animations
  local animation = animations and animations[animationIndex + 1]
  if not animation then
    return 0 / 0, 0 / 0
  end
  local min = math.huge
  local max = -math.huge
  local seen = {}
  for _, channel in ipairs(animation.channels or {}) do
    seen[channel.sampler] = true
  end
  for samplerIndex, _ in pairs(seen) do
    local sampler = (animation.samplers or {})[samplerIndex + 1]
    local input = sampler and (host.gltf.accessors or {})[(sampler.input or -1) + 1]
    if input then
      if type(input.min) == "table" and input.min[1] ~= nil then
        min = math.min(min, input.min[1])
      end
      if type(input.max) == "table" and input.max[1] ~= nil then
        max = math.max(max, input.max[1])
      end
    end
  end
  local function finite(x) return x == x and x ~= math.huge and x ~= -math.huge end
  if not finite(min) or not finite(max) then
    return 0 / 0, 0 / 0
  end
  return min, max
end

-- Maps a requested timestamp on the infinite timeline to the effective
-- timestamp within the animation data range [0, T] (KHR_interactivity §4.6).
function anim_effectiveAnimationTime(requested, maxTime)
  if maxTime == 0 or maxTime ~= maxTime then
    return 0
  end
  local s
  if requested > 0 then
    s = math.ceil((requested - maxTime) / maxTime)
  else
    s = math.floor(requested / maxTime)
  end
  return requested - s * maxTime
end

-- Samples animation `animationIndex` at virtual-timeline position
-- `requested` and writes the resulting channel values straight into
-- `host.gltf`'s node objects. `onChannelWrite(pointer, value)` is called for
-- each written channel. Returns nil if the animation doesn't exist, else a
-- table `{ t = <effective sample time>, wroteAny = <bool> }`.
function anim_applyAnimationAt(host, animationIndex, requested, onChannelWrite)
  local animations = host.gltf and host.gltf.animations
  local animation = animations and animations[animationIndex + 1]
  if not animation then
    return nil
  end
  local _, maxTime = anim_getAnimationTimeRange(host, animationIndex)
  if maxTime ~= maxTime then maxTime = 0 end
  local t = anim_effectiveAnimationTime(requested, maxTime)
  local wroteAny = false
  for _, channel in ipairs(animation.channels or {}) do
    local sampler = (animation.samplers or {})[channel.sampler + 1]
    local target = channel.target
    if sampler and target and target.node ~= nil then
      local times = anim_readAccessorElements(host, sampler.input)
      local values = anim_readAccessorElements(host, sampler.output)
      if times and values and #times > 0 then
        local timesScalar = {}
        for i = 1, #times do timesScalar[i] = times[i][1] end
        local interpolation = sampler.interpolation or "LINEAR"
        local stride = (interpolation == "CUBICSPLINE") and 3 or 1
        local cubicOffset = (stride == 3) and 1 or 0
        local sampled
        if t <= timesScalar[1] then
          sampled = values[0 * stride + cubicOffset + 1]
        elseif t >= timesScalar[#timesScalar] then
          sampled = values[(#timesScalar - 1) * stride + cubicOffset + 1]
        else
          local k = 0
          while k + 2 <= #timesScalar and timesScalar[k + 2] < t do
            k = k + 1
          end
          local t0 = timesScalar[k + 1]
          local t1 = timesScalar[k + 2]
          local u = (t1 > t0) and ((t - t0) / (t1 - t0)) or 0
          local v0 = values[k * stride + cubicOffset + 1]
          local v1 = values[(k + 1) * stride + cubicOffset + 1]
          if interpolation == "STEP" then
            sampled = v0
          elseif target.path == "rotation" then
            sampled = kmath_quatSlerp(v0, v1, u)
          else
            sampled = {}
            for i = 1, #v0 do sampled[i] = v0[i] + (v1[i] - v0[i]) * u end
          end
        end
        local node = (host.gltf.nodes or {})[target.node + 1]
        if node then
          if target.path == "weights" then
            node.weights = sampled
          else
            node[target.path] = sampled
          end
          if onChannelWrite then
            onChannelWrite("/nodes/" .. target.node .. "/" .. target.path, sampled)
          end
          wroteAny = true
        end
      end
    end
  end
  return { t = t, wroteAny = wroteAny }
end
